File: java/src/test/java/com/bc/libwally/Bip39Test.java
package com.bc.libwally;

import com.bc.libwally.bip39.Bip39Entropy;
import com.bc.libwally.bip39.Bip39Exception;
import com.bc.libwally.bip39.Bip39Mnemonic;
import com.bc.libwally.bip39.Bip39Seed;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Arrays;

import static com.bc.libwally.util.TestUtils.assertThrows;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(JUnit4.class)
public class Bip39Test {

    private static final String validMnemonic = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";
    private static final String validMnemonic24 = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon art";

    @Test
    public void testGetWordList() {
        String[] bip39Words = Bip39Mnemonic.getBip39Words();
        assertEquals(2048, bip39Words.length);
        assertEquals("abandon", bip39Words[0]);
    }

    @Test
    public void testMnemonicIsValid() {
        assertTrue(Bip39Mnemonic.isValid(validMnemonic));
        assertFalse(Bip39Mnemonic.isValid("notavalidword"));
        assertFalse(Bip39Mnemonic.isValid("abandon"));
        assertFalse(Bip39Mnemonic.isValid(new String[]{"abandon", "abandon"}));
    }

    @Test
    public void testInitializeMnemonic() {
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(validMnemonic);
        assertTrue(Arrays.deepEquals(validMnemonic.split(" "), mnemonic.getWords()));
    }

    @Test
    public void testInitializeMnemonicFromBytes() {
        byte[] bytes = new byte[32];
        Bip39Entropy entropy = new Bip39Entropy(bytes);
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(entropy);
        assertTrue(Arrays.deepEquals(validMnemonic24.split(" "), mnemonic.getWords()));
    }

    @Test
    public void testInitializeInvalidMnemonic() {
        assertThrows("",
                     Bip39Exception.class,
                     () -> new Bip39Mnemonic(new String[]{"notavalidword"}));
    }

    @Test
    public void testMnemonicLosslessStringConvertible() {
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(validMnemonic);
        assertEquals(validMnemonic, mnemonic.getMnemonic());
    }

    @Test
    public void testMnemonicToEntropy() {
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(validMnemonic);
        assertEquals("00000000000000000000000000000000", mnemonic.getEntropy().getHexData());
        Bip39Mnemonic mnemonic2 = new Bip39Mnemonic(
                "legal winner thank year wave sausage worth useful legal winner thank yellow");
        assertEquals("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f", mnemonic2.getEntropy().getHexData());
    }

    @Test
    public void testEntropyToMnemonic() {
        Bip39Mnemonic expectedMnemonic = new Bip39Mnemonic(
                "legal winner thank year wave sausage worth useful legal winner thank yellow");
        Bip39Entropy entropy = new Bip39Entropy("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f");
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(entropy);
        assertEquals(expectedMnemonic, mnemonic);
    }

    @Test
    public void testEntropyLosslessStringConvertible() {
        Bip39Entropy entropy = new Bip39Entropy("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f");
        assertEquals("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f", entropy.getHexData());
    }

    @Test
    public void testMnemonicToSeedHexString() {
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(validMnemonic);
        assertEquals(
                "c55257c360c07c72029aebc1b53c05ed0362ada38ead3e3e9efa3708e53495531f09a6987599d18264c1e1c92f2cf141630c7a3c4ab7c81b2f001698e7463b04",
                mnemonic.getSeedHex("TREZOR").getHexData());
        assertEquals(
                "5eb00bbddcf069084889a8ab9155568165f5c453ccb85e70811aaed6f6da5fc19a5ac40b389cd370d086206dec8aa6c43daea6690f20ad3d8d48b2d2ce9e38e4",
                mnemonic.getSeedHex().getHexData());
        assertEquals(
                "5eb00bbddcf069084889a8ab9155568165f5c453ccb85e70811aaed6f6da5fc19a5ac40b389cd370d086206dec8aa6c43daea6690f20ad3d8d48b2d2ce9e38e4",
                mnemonic.getSeedHex("").getHexData());
    }

    @Test
    public void testSeedLosslessStringConvertible() {
        Bip39Mnemonic mnemonic = new Bip39Mnemonic(validMnemonic);
        Bip39Seed expectedSeed = mnemonic.getSeedHex("TREZOR");
        Bip39Seed actualSeed = new Bip39Seed(
                "c55257c360c07c72029aebc1b53c05ed0362ada38ead3e3e9efa3708e53495531f09a6987599d18264c1e1c92f2cf141630c7a3c4ab7c81b2f001698e7463b04");
        assertEquals(expectedSeed, actualSeed);
    }

}

File: android/app/src/androidTest/java/com/bc/libwally/Bip39Test.kt
package com.bc.libwally

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.bc.libwally.bip39.Bip39Entropy
import com.bc.libwally.bip39.Bip39Exception
import com.bc.libwally.bip39.Bip39Mnemonic
import com.bc.libwally.bip39.Bip39Seed
import com.bc.libwally.util.assertThrows
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

@RunWith(AndroidJUnit4::class)
class Bip39Test {

    private val validMnemonic =
        "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about"
    private val validMnemonic24 =
        "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon art"

    @Test
    fun testGetWordList() {
        val bip39Words = Bip39Mnemonic.getBip39Words()
        assertEquals(2048, bip39Words.size.toLong())
        assertEquals("abandon", bip39Words[0])
    }

    @Test
    fun testMnemonicIsValid() {
        assertTrue(Bip39Mnemonic.isValid(validMnemonic))
        assertFalse(Bip39Mnemonic.isValid("notavalidword"))
        assertFalse(Bip39Mnemonic.isValid("abandon"))
        assertFalse(Bip39Mnemonic.isValid(arrayOf("abandon", "abandon")))
    }

    @Test
    fun testInitializeMnemonic() {
        val mnemonic = Bip39Mnemonic(validMnemonic)
        assertTrue(Arrays.deepEquals(validMnemonic.split(" ").toTypedArray(), mnemonic.words))
    }

    @Test
    fun testInitializeMnemonicFromBytes() {
        val bytes = ByteArray(32)
        val entropy = Bip39Entropy(bytes)
        val mnemonic = Bip39Mnemonic(entropy)
        assertTrue(Arrays.deepEquals(validMnemonic24.split(" ").toTypedArray(), mnemonic.words))
    }

    @Test
    fun testInitializeInvalidMnemonic() {
        assertThrows<Bip39Exception>("") { Bip39Mnemonic(arrayOf("notavalidword")) }
    }

    @Test
    fun testMnemonicLosslessStringConvertible() {
        val mnemonic = Bip39Mnemonic(validMnemonic)
        assertEquals(validMnemonic, mnemonic.mnemonic)
    }

    @Test
    fun testMnemonicToEntropy() {
        val mnemonic = Bip39Mnemonic(validMnemonic)
        assertEquals("00000000000000000000000000000000", mnemonic.entropy.hexData)
        val mnemonic2 =
            Bip39Mnemonic("legal winner thank year wave sausage worth useful legal winner thank yellow")
        assertEquals("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f", mnemonic2.entropy.hexData)
    }

    @Test
    fun testEntropyToMnemonic() {
        val expectedMnemonic =
            Bip39Mnemonic("legal winner thank year wave sausage worth useful legal winner thank yellow")
        val entropy = Bip39Entropy("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f")
        val mnemonic = Bip39Mnemonic(entropy)
        assertEquals(expectedMnemonic, mnemonic)
    }

    @Test
    fun testEntropyLosslessStringConvertible() {
        val entropy = Bip39Entropy("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f")
        assertEquals("7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f7f", entropy.hexData)
    }

    @Test
    fun testMnemonicToSeedHexString() {
        val mnemonic = Bip39Mnemonic(validMnemonic)
        assertEquals(
            "c55257c360c07c72029aebc1b53c05ed0362ada38ead3e3e9efa3708e53495531f09a6987599d18264c1e1c92f2cf141630c7a3c4ab7c81b2f001698e7463b04",
            mnemonic.getSeedHex("TREZOR").hexData
        )
        assertEquals(
            "5eb00bbddcf069084889a8ab9155568165f5c453ccb85e70811aaed6f6da5fc19a5ac40b389cd370d086206dec8aa6c43daea6690f20ad3d8d48b2d2ce9e38e4",
            mnemonic.seedHex.hexData
        )
        assertEquals(
            "5eb00bbddcf069084889a8ab9155568165f5c453ccb85e70811aaed6f6da5fc19a5ac40b389cd370d086206dec8aa6c43daea6690f20ad3d8d48b2d2ce9e38e4",
            mnemonic.getSeedHex("").hexData
        )
    }

    @Test
    fun testSeedLosslessStringConvertible() {
        val mnemonic = Bip39Mnemonic(validMnemonic)
        val expectedSeed: Bip39Seed = mnemonic.getSeedHex("TREZOR")
        val actualSeed = Bip39Seed(
            "c55257c360c07c72029aebc1b53c05ed0362ada38ead3e3e9efa3708e53495531f09a6987599d18264c1e1c92f2cf141630c7a3c4ab7c81b2f001698e7463b04"
        )
        assertEquals(expectedSeed, actualSeed)
    }
}